### PR TITLE
ICurve implemented

### LINKED ISSUE #383: Transform a group of curves.
# Expected Behavior

Create a mixed collection of curves.
 ```C#
 crvs = new List<NurbsBase>{arc, circle, line, polyline, polycurve, nurbs };
```

Apply a transformation to all curves in the collection.

```C#
var xForm = Transform.Translation(new Point3(0,0,0), new Point3(10,0,0));
var xFormCrvs = crvs.Select(c => c.Transform(xForm)).ToList();
```
# Issue
NurbsBase does not implement a Transform method. Requires reflection or a switch to cast to appropriate type.

# Suggested Solution

1. Implement Transform on NurbsBase and override as required
2. Implement new ICurve interface to implement ITransformable and IEquatable<T>. NurbsBase implements ICurve.

### PR DESCRIPTION
**Pull request by NishantC5**
### What type of PR is this? (check all applicable)

- [x] ♻️ Refactor
- [ ] ✨ Feature
- [ ] 🐛 Bug Fix
- [ ] 👷 Optimization
- [ ] 📝 Documentation Update
- [ ] 🔖 Release
- [ ] 🚩 Other

### Description

This PR introduces ICurve interface in order to implement Transform function in NurbsBase. With these changes, one can
make a `List<NurbsBase> geometries` which can contain any transformable geometry and `geometries.Select(c => c.Transform(xForm)).ToList()` would execute the `Trasnform` function of each respective geometry.

### Related Tickets & Documents

Closes #383

### Added tests?

- [x] 👍 yes
- [ ] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

### Added to documentation?

- [ ] 📓 docs
- [x] 🙅 no documentation needed

### CODE REVIEW

**Comment by d3ssy** [comment]
Thank you @NishantC5! Great work. 

I have some comments. I think it's fine to have ICurve implemented by all curve-like geometries, but it doesn't seem logical for Plane or Surface to implement ICurve.

The other classes can implement ITransformable on their own.

Alternatively, we can define an IGeometry interface that implements IEquatable and ITransformable and is itself implemented by all geometric objects in the library. 

@sonomirco, @cesarecaoduro any thoughts?

**Comment by cesarecaoduro** [comment]
I recon having an IGeometry will make more sense. Similar to BHoM, we can also implement a deeper level like IObject that is the atomic interface for everything.

Sent from Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by sonomirco** [comment]
Thanks, @NishantC5, for the work!
I agree @d3ssy. It is better to introduce `IGeometry` instead of having `Plane` implementing `ICurve`.

File: src/GShark/Geometry/NurbsBase.cs [hunk 1]
```diff
@@ -853,5 +854,11 @@ public override int GetHashCode()
 
             return sBldr.ToString().GetHashCode();
         }
+
+        public NurbsBase Transform(TransformMatrix t)
```
Comment (sonomirco) [inline src/GShark/Geometry/NurbsBase.cs:858]: @NishantC5, Could you please document this function? Thanks.

**Comment by NishantC5** [comment]
Ah yes! It doesn't make sense for plane and surface to implement ICurve! As per the above discussion, I will introduce IGeometry which implements IEquatable and ITransformable and this would be implemented by all the geometry objects. Thanks for the feedback @d3ssy @cesarecaoduro and @sonomirco.

File: src/GShark/Geometry/NurbsBase.cs (hunk 1 above)
Comment (NishantC5) [reply src/GShark/Geometry/NurbsBase.cs:858]: Yes sure, I'll get back with the documented code in the next commit.

**Comment by sonomirco** [comment]
@NishantC5 can we merge this PR?

**Comment by NishantC5** [comment]
@sonomirco yes.

**Pull request merged**